vllm/model_executor/layers/sampler.py

```diff
@@ -149,45 +172,52 @@ def _apply_penalties(
     output_tokens: List[List[int]],
     presence_penalties: List[float],
     frequency_penalties: List[float],
-    vocab_size: int,
 ) -> torch.Tensor:
-    num_seqs = logits.shape[0]
-    # Collect the indices of sequences that have non-zero penalties.
-    indices = []
+    num_seqs, vocab_size = logits.shape
+    can_skip = True
     for i in range(num_seqs):
         if not output_tokens[i]:
             continue
         p = presence_penalties[i]
         f = frequency_penalties[i]
         if abs(p) < _SAMPLING_EPS and abs(f) < _SAMPLING_EPS:
             continue
-        indices.append(i)
+        can_skip = False
+        break
 
     # Return early if all sequences have zero penalties.
-    if not indices:
+    if can_skip:
         return logits
```

Comment:
style nit: I think you can use `for/else` statement instead, but it's up to you.